File: python/lesson-11-sqlalchemy/exercises/exercise.py
"""Lesson 11 Exercise — Build a Product Catalog with SQLAlchemy 2.0 async."""

import asyncio

# TODO: Set up the async engine and session factory for SQLite:
#   DATABASE_URL = "sqlite+aiosqlite:///./products.db"
# TODO: Create a DeclarativeBase

# TODO: Define a Product ORM model with:
#   id (PK), name (str, unique), price (float), stock (int, default=0),
#   created_at (DateTime, server_default=func.now())

# TODO: Create an init_db() coroutine that creates all tables

# TODO: Implement a ProductRepository class with:
#   - get_all() -> list[Product]
#   - get_by_id(id) -> Product | None
#   - get_by_name(name) -> Product | None
#   - create(name, price, stock=0) -> Product
#   - update(product, name=None, price=None, stock=None) -> Product
#   - delete(product) -> None

# TODO: Write a demo() coroutine that:
#   1. Initializes the DB
#   2. Creates 3 products
#   3. Lists all products
#   4. Updates the price of one product
#   5. Deletes one product
#   6. Lists remaining products

async def demo() -> None:
    # TODO: implement
    pass


if __name__ == "__main__":
    asyncio.run(demo())

File: dotnet/lesson-11-efcore/exercises/Exercise.cs
// ── Lesson 11 Exercise ────────────────────────────────────────────────────────
// Build a Product Catalog API with EF Core + SQLite.

using Microsoft.EntityFrameworkCore;

var builder = WebApplication.CreateBuilder(args);

// TODO: Register AppDbContext with SQLite ("Data Source=products.db")
// TODO: Register IProductRepository as scoped
builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

var app = builder.Build();

// TODO: Apply migrations on startup

app.UseSwagger();
app.UseSwaggerUI();

// TODO: Map routes for Product CRUD under /api/products
// GET    /api/products         → list all
// GET    /api/products/{id}    → get one or 404
// POST   /api/products         → create (name + price required)
// PUT    /api/products/{id}    → update name/price/stock (partial)
// DELETE /api/products/{id}    → delete or 404

app.Run("http://localhost:5001");

// ── Models ────────────────────────────────────────────────────────────────────

// TODO: Define a Product entity with: Id, Name (required), Price (decimal),
//       Stock (int, default 0), CreatedAt (DateTime, auto)
// TODO: Define CreateProductDto and UpdateProductDto records
// TODO: Define ProductDto record for API responses

// ── DbContext ─────────────────────────────────────────────────────────────────

// TODO: Create AppDbContext with DbSet<Product>
//       Add a unique index on Name in OnModelCreating

// ── Repository ────────────────────────────────────────────────────────────────

// TODO: Define IProductRepository interface
// TODO: Implement ProductRepository using AppDbContext
//       All methods must be async
